synapse/storage/_base.py

```diff
@@ -1000,7 +1001,8 @@ def _invalidate_cache_and_stream(self, txn, cache_func, keys):
             # __exit__ called after the transaction finishes.
             ctx = self._cache_id_gen.get_next()
             stream_id = ctx.__enter__()
-            txn.call_finally(ctx.__exit__, None, None, None)
+            txn.call_on_exception(ctx.__exit__, None, None, None)
+            txn.call_after(ctx.__exit__, None, None, None)
```

Comment:
could maybe do with a comment here about why this is a useful thing to do (ie, make sure we close the context before calling the other thing)